### PR TITLE
Upgrade React 16.3.2 to React 16.4.1

### PR DESCRIPTION
**Pull request by abotteram**
## Description

This PR updates react from version 16.3.2 to version 16.4.1.

Reason for this PR is because I would like to use `react-test-renderer` for testing components that use React's new `forwardRef` feature, which is not supported by enzyme. As you can read in the [changelog for React 16.4](https://reactjs.org/blog/2018/05/23/react-v-16-4.html), support for using `react-test-renderer` with `forwardRef` was added in this version.

Not being able to test components with `forwardRef` is currently blocking https://github.com/WordPress/gutenberg/pull/6261.

Since this is a minor version bump no breaking changes are to be expected, and while testing this I did not encounter any unexpected behaviour.

## How has this been tested?

The unit tests still work.

## Types of changes

New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -44,12 +44,12 @@
 		"prop-types": "15.5.10",
 		"querystringify": "1.0.0",
 		"re-resizable": "4.4.8",
-		"react": "16.3.2",
+		"react": "^16.4.1",
```
Comment (tofumatt) [inline package.json]: Why the change to using caret? Locking down package versions as tightly as possible is a good idea, so I think this should be changed to `16.4.1` over `^16.4.1`.

File: package.json [hunk 2]
```diff
@@ -44,12 +44,12 @@
...
 		"react-autosize-textarea": "3.0.2",
 		"react-click-outside": "2.3.1",
 		"react-color": "2.13.4",
 		"react-datepicker": "1.4.1",
-		"react-dom": "16.3.2",
+		"react-dom": "^16.4.1",
```
Comment (tofumatt) [inline package.json]: Same here re: caret ranges.

**Review by tofumatt — DISMISSED** [review]
I don't think we should be specifying version ranges with the caret.

File: packages/element/package-lock.json [hunk 3]
```diff
@@ -0,0 +1,146 @@
+{
```
Comment (gziolo) [inline packages/element/package-lock.json]: This file should be removed. We don't want to maintain package lock files for individual packages. It's a bug with Lerna that it doesn't respect `.npmrc` file ... I observe that it happens where you run `npm install` when there are 2 different versions of the package in two `package.json` files.

**Review by gziolo — CHANGES_REQUESTED** [review]
There are a few more occurrences of old version of React in php files. We need to update them before we proceed.

**Review by gziolo — APPROVED** [review]
Code looks good, let me give it a spin locally.

**Comment by gziolo** [comment]
Sidenote: we probably should start using `react-test-renderer` more instead of `Enzyme`, because of all the issues `Enzyme` caused by not keeping up to date with React releases.
